src/TLuaInterpreter.cpp

```diff
@@ -14281,7 +14281,20 @@ void TLuaInterpreter::parseJSON(QString& key, const QString& string_data, const
         if (match.capturedStart() != -1) {
             QString title = match.captured(1);
             QString initialText = match.captured(2);
-            initialText.replace(QStringLiteral(R"(\n)"), QStringLiteral("\n"));
+            int j=0;
+            while((j = initialText.indexOf("\\", j)) != -1){
+                // while we find a backslash, look at the next character, decide what to replace them both with
+                switch(initialText.at(j+1).unicode()){
+                    case 'n' : initialText.replace(j,2,QLatin1String("\n")); break;
+                    case 't' : initialText.replace(j,2,QLatin1String("\t")); break; // IRE instead sends it out as \u0009
+                    case '\"' : initialText.replace(j,2,QLatin1String("\"")); break;
+                    case '\\' : initialText.replace(j,2,QLatin1String("\\")); break;
+                    case 'u' : // if Achaea sends any \u codes other than tab then this could be more generic
+                        if(initialText.midRef(j,6) == "\\u0009") initialText.replace(j,6,QLatin1String("\t"));
```

Comment:
:information_source: No single line `if (...) {...}`s please - and don't skip on the braces `{` ... `}` either - `clang-format` might take care of this if you run it over your changes (you can select them and then run it from within Qt Creator {`Tools` -> `Beautifier` -> `Clang-Format` -> `Format at Cursor`} if you set it to use the `Predefined style` -> `file` (which we ship within the source files as `./src.clang-format`)...

:thinking: Actually though, it strikes me that the original `QString::replace(...)` may have been correct in a way that it seems to *me* *might*(?) be being reproduced incorrectly in the replacement code. Or I could have gotten it **wrong** anyhow. Specifically I think it comes down to the fact that the C++11 raw string literal `R"(\n)"` is literally the backslash character followed by a lower case `n` - in the absence of the raw string literal this would be represented as a old-school C string as the TWO characters `"\\n"` which the original code is fixing to be the SINGLE character `'\n'`.

Reply:
I think I improved things.  For some reason I was under the impression that those functions needed strings, but they work with characters too, so I switched most of those to characters.  Also the if statement is more readable now.

> Actually though, it strikes me that the original QString::replace(...) may have been correct in a way that it seems to me might(?) be being reproduced incorrectly

I didn't want to reproduce the current behavior.  Let's say you have a message in the game that has the two characters `\n`.  With the original code, when you edit your journal or whatever, the game would send a GMCP message that contains `\\n`, and when it arrives, the client would put a backslash and newline in the editor window instead of a backslash and letter `n`.

I think using the replace function to replace every `\\` and then every `\n` is likely very efficient compared to a while loop and whatever it costs to replace parts one at a time and resize the string.  However, if I went with `\\` -> `\` and then `\n` -> `(newline)` then an input like `\\n` would first turn into `\n` and then a newline character.  Doing it the opposite order would end up with a backslash and newline.  So I think just a simple replace for all backslashes and then newlines will only work for escaping, while unescaping would need to happen in one pass like this.  It's likely possible to make it more efficient though.

Reply:
:thought_balloon: Of course what was getting my brain screwed up was considering at what point the representation of a backslash '\\' followed by, say, a lower case 'n' gets converted into the byte with the value \x0A i.e the Line-Feed character. I think the *escaping* has temporarily escaped me. :slightly_frowning_face: